NVorbis/Ogg/ContainerReader.cs

```diff
@@ -0,0 +1,154 @@
+﻿using NVorbis.Contracts;
+using NVorbis.Contracts.Ogg;
+using System;
+using System.Collections.Generic;
+using System.IO;
+
+namespace NVorbis.Ogg
+{
+    /// <summary>
+    /// Implements <see cref="Contracts.IContainerReader"/> for Ogg format files for low memory cost.
+    /// </summary>
+    public sealed class ContainerReader : Contracts.IContainerReader
+    {
+        internal static Func<Stream, bool, Func<Contracts.IPacketProvider, bool>, IPageReader> CreatePageReader { get; set; } = (s, cod, cb) => new PageReader(s, cod, cb);
+        internal static Func<Stream, bool, Func<Contracts.IPacketProvider, bool>, IPageReader> CreateForwardOnlyPageReader { get; set; } = (s, cod, cb) => new ForwardOnlyPageReader(s, cod, cb);
+
+        private IPageReader _reader;
+        private List<WeakReference<Contracts.IPacketProvider>> _packetProviders;
+        private bool _foundStream;
+
+        /// <summary>
+        /// Gets or sets the callback to invoke when a new stream is encountered in the container.
+        /// </summary>
+        public NewStreamHandler NewStreamCallback { get; set; }
+
+        /// <summary>
+        /// Gets a list of streams available from this container.
+        /// </summary>
+        public IReadOnlyList<Contracts.IPacketProvider> Streams
```

Comment:
This property would be much better as a method in my opinion since it allocates a new list every time. I don't know if anyone using the library will be using this class, but it is public.

Reply:
Hmmm...  That's really an implementation detail, so I'm hesitant to change the public API there.  On the other hand, returning a new list every time is not a great idea...
What about a custom IReadOnlyList<IPacketProvider> implementation?  Use that instead of the List<WeakReference<IPacketProvider>> instance it uses now?

Reply:
My main concern was that it being a property would hide the allocation. It would be easy to access the property multiple times when working with the list. A well named method would better communicate that you should just call it once and assign the created list to a variable. I'm not sure how often this property is accessed but this could probably be left alone and it wouldn't matter much at all even if someone misused the property. I was just going to suggest deprecating the property and saying to use the method instead.

Reply:
Well, I looked at the possibility of implementing a WeakList<T> that implements IReadOnlyList<T>...  Nope, not in this release!  Too many corner cases and "gotchas" for this stage in the release cycle.  So I think I will change the public API to be a method.  Thanks for the suggestion.